### ISSUE TITLE
SvelteKit goto doesn't re-render state when navigating to a page in the same directory

### ISSUE THREAD

**Issue by Beyondo** [sveltejs/kit#11116]
### Describe the bug

Sveltekit does not react to goto navigations properly in a special case.

### Reproduction

When I `goto('/users/user-[UID_SLUG]')`, it does not change the state reactively. But when I go from `'/users/user-[UID_SLUG]'` to `/`, it changes the state reactively.

The two workarounds I have are either using an anchor link with `data-sveltekit-reload` (non-reactive) or setting the state manually when the user clicks the anchor link which is kinda frustrating.

But what I've noticed in the 2nd workaround, is that I don't just set `$currentUser = {...};`, no, I actually have to do this:
```
$currentUser = null;
setTimeout(() => {
    $currentUser = { ... };
}, 0);
```
Something weird is going on and I honestly cannot figure it out.

### System Info

```Shell
System:
    OS: Windows 10 10.0.22621
    CPU: (16) x64 AMD Ryzen 7 5800H with Radeon Graphics
    Memory: 6.01 GB / 31.86 GB
  Binaries:
    Node: 19.5.0 - C:\Program Files\nodejs\node.EXE
    npm: 9.3.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (119.0.2151.58)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @sveltejs/adapter-cloudflare: ^2.3.3 => 2.3.3
    @sveltejs/adapter-node: ^1.3.1 => 1.3.1
    @sveltejs/adapter-static: ^2.0.2 => 2.0.2
    @sveltejs/kit: ^1.20.4 => 1.22.3
    svelte: ^4.0.5 => 4.1.1
    vite: ^4.4.2 => 4.4.7
```

### Severity

serious, but I can work around it

**Comment by Conduitry** [comment]
Please provide a minimal reproduction in the form of a a link to a repository on GitHub, as requested in the issue template.

**Comment by eltigerchino** [comment]
https://kit.svelte.dev/docs/state-management#component-state-is-preserved

**Comment by PH4NTOMiki** [comment]
this is really buggy, especially with Svelte 5

**Comment by mikekok** [comment]
> this is really buggy, especially with Svelte 5

Exactly.
When using Svelte 4 syntax in Svelte 5, I can't get data to reload reactively withing the same route.

**Comment by OrakMoya** [comment]
Also experiencing this, though I did find a workaround that makes a bit more sense?
```svelte
<script>
	let { data } = $props(); // Data returned by +page.server.js
	let currentUser = $derived(data.currentUser);
</script>
```
`data` reacts to changes returned by the server for the new route, so I added the $derived to make Svelte recalculate the currentUser.

**Comment by alanxp** [comment]
> Also experiencing this, though I did find a workaround that makes a bit more sense?
> 
> <script>
> 	let { data } = $props(); // Data returned by +page.server.js
> 	let currentUser = $derived(data.currentUser);
> </script>
> `data` reacts to changes returned by the server for the new route, so I added the $derived to make Svelte recalculate the currentUser.
> 
> Edit: just saw the docs link, that does exactly this. This is really counter-intuitive and not what I expect svelte to do after navigating to a new page.
> 
> I should add that this method doesnt work when the page has to change that data reactively. Instead of using $derived, I need to do this:
> 
> <script>
> 	let { data } = $props(); // Data returned by +page.server.js
> 	let currentUser = $state(data.currentUser);
> 	$effect(() => {
> 		currentUser = data.currentUser;
> 	});
> </script>

yep, i can confirm that doing this is necesary, which is not ideal

**Comment by ColeyG** [comment]
Just lost a few hours to this, maybe consider documenting?

**Comment by NeroBlackstone** [comment]
> <script>
> 	let { data } = $props(); // Data returned by +page.server.js
> 	let currentUser = $state(data.currentUser);
> 	$effect(() => {
> 		currentUser = data.currentUser;
> 	});
> </script>

Not work for me.. Is this caused by using a library written in svelte 4 in a svelte5 project?

**Comment by NeroBlackstone** [comment]
> Not work for me.. Is this caused by using a library written in svelte 4 in a svelte5 project?

Yes, I think It's reason.

**Comment by winteryoung** [comment]
> [@eltigerchino](https://github.com/eltigerchino) This issue should be reopened. The official doc still poses a reactivity problem. With $derived, the page is able to refresh on new data from load. But **lost** the reactivity. I think [@OrakMoya](https://github.com/OrakMoya) 's edited answer hit the point.
> 
> [@OrakMoya](https://github.com/OrakMoya) It just cost me several hours. Thank you for your solution. A little different from yours, I changed $state to $derived, and this works for me:
> 
> ```
> <script lang="ts">
>   import { page } from "$app/state";
>   import MonthlyBill from "$lib/components/monthly-bill/monthly-bill.svelte";
> 
>   const data = $derived(page.data);
> 
>   let apiRet = $derived(data.ret);
>   $effect(() => {
>     apiRet = data.ret;
>   });
> </script>
> 
> <MonthlyBill {apiRet} />
> ```
> 
> > Also experiencing this, though I did find a workaround that makes a bit more sense?
> > <script>
> > 	let { data } = $props(); // Data returned by +page.server.js
> > 	let currentUser = $derived(data.currentUser);
> > </script>
> > `data` reacts to changes returned by the server for the new route, so I added the $derived to make Svelte recalculate the currentUser.
> > Edit: just saw the docs link, that does exactly this. This is really counter-intuitive and not what I expect svelte to do after navigating to a new page.
> > I should add that this method doesnt work when the page has to change that data reactively. Instead of using $derived, I need to do this:
> > <script>
> > 	let { data } = $props(); // Data returned by +page.server.js
> > 	let currentUser = $state(data.currentUser);
> > 	$effect(() => {
> > 		currentUser = data.currentUser;
> > 	});
> > </script>

Also, a by-product of this solution is `apiRet` in the sub component got evaluated twice on page loading. This doesn't happen if I changed `apiRet` to `$state`.

**Comment by JeffWScott** [comment]
Should be reopened.

I'm just using a +page.ts and +page.svelte, very simple setup, just passing in the [id] param via the data object.

When I click on an anchor tag from /foo/1 with a link to /foo/2 I can see that the [id] param changes in the +page.ts but the data object in the +page.svelte never retriggers and all the data on the page is the same.

The only thing that works for me is to run my a tags into buttons and call this:
```
window.location.assign(`/foo/${id}`)
```

even trying to run invalidateAll with goto will not work, I need to do the location assignment.

**Comment by teemingc** [comment]
> Should be reopened.
> 
> I'm just using a +page.ts and +page.svelte, very simple setup, just passing in the [id] param via the data object.
> 
> When I click on an anchor tag from /foo/1 with a link to /foo/2 I can see that the [id] param changes in the +page.ts but the data object in the +page.svelte never retriggers and all the data on the page is the same.
> 
> The only thing that works for me is to run my a tags into buttons and call this:
> 
> ```
> window.location.assign(`/foo/${id}`)
> ```
> 
> even trying to run invalidateAll with goto will not work, I need to do the location assignment.

Can you share a minimal reproduction? (repository)

**Comment by Beyondo** [comment]
I don't exactly remember (or even sure if I completely understand) what was my problem in this issue here was as it has been a while (~2 years).

But now that I'm better at Svelte, while reading this I realized there slight confusion from my younger self if it's what I think it is. 

Basically, `goto` is not the problem. Your code logic is.

For example, you could've had code like this.
`routes/thing/[id]/+page.svelte`:
```ts
onMount(() => {
  const current_thing_id = (new URL(window.location.href)).pathname.match(/^\/thing\/([^/]+)/)[1]; // Yup...
  /* load thing */
  active_thing = thing;
});
```

But unlike navigating from `/home/+page.svelte` to `/about/+page.svelte` (the kind of pages everyone is used to create), **navigating between [slugs] like `[id]` will NOT re-trigger `onMount`.**

- Navigating from `page1/+page.svelte` to another `page2/+page.svelte` will run the `onMount` of each and their whole `<script>`.
- BUT, navigating from `thing/[id]/+page.svelte (id=1)` to `thing/[id]/+page.svelte (id=2)` is still in the **same `+page.svelte`** (even if it seems like a different page in your browser; e.g. `/thing/1` -> `/thing/2`) so not a single line of your `+page.svelte` will ever actually be re-ran again.

And _that's the confusing_ part (at least I believe so). We know that navigation = different pages in browser, and because that also _mostly_ meant different `+page.svelte` too, many devs might've subconsciously got confused when they started using slugs like `[id]` that they should also re-trigger `onMount`, but the thing is: `onMount` is just for the component itself. It will never re-trigger in Svelte's `goto` if you are simply navigating the `[slug]` part while _still inside it._

And when you navigated manually with **non-svelte methods (non-reactive)**, this literally just refreshed the page which of course forced the `onMount` to re-trigger (which ran your code 'normally' but not reactively), creating an even bigger misunderstanding that `goto` was the problem but it never was.

So anyways, how do you detect navigations in the same `+page.svelte` then?

## The Correct Way To Load Data for Svelte (CSR)

You must use the universal `load` function of `+page.ts`. And just by referencing your `id` parameter inside that function, it'd be added as "dependency," and thus `load` will keep re-running every time that parameter changes, without re-mounting.

Example `page.ts:`
```ts
export const prerender = false;
export const ssr = false;

export async function load(event) {
    await event.parent();
    const thing_id = event.params.id; // Yes, no need to do mind-numbing regex now for CSR.
}
```
This will first run on page load. Plus **re-run** every time `event.params.id` changes. Exactly what you need. Also `await event.parent()` is a good practice if your `+page.ts`'s `load` logic depends on prior `+layout.ts`'s `load`.

Here's a better demo code:
```ts
import { app } from "$lib/app.svelte";
import { Toast } from "$lib/toast";
import { redirect } from "@sveltejs/kit";

export const prerender = false;
export const ssr = false;

export async function load(event) {
    await event.parent();
    try {
        const thing_id = event.params.id;
        const thing = await app.thing_service.get(thing_id);
        if (!thing) {
            Toast.make_text("Thing not found. Moving to Home...");
            throw redirect(302, "/");
        }
        return { thing };
    } catch {
        throw redirect(302, "/");
    }
}
```
Note that the `redirect()` here is fully client-side (CSR), there's no actual HTTP responses or HTTP codes sent at all. Basically, it's exactly like using `goto()`. But difference is that using **`redirect()` gives you type safety guarantees** in Svelte, while **`await goto()` does not.** Basically, without it, your `+page.svelte` (if you're using `lang="ts"`) would infer the type of the `$props()` as `{ data: Thing | undefined }` but we want to guarentee it to be `{ data: Thing }` and that's why we use `throw redirect()`.

Now the question is, how do you "react" to `data` changes? Or in other words, to `load()` being re-triggered?

Answer: You don't need to. When you pull the prop `data` in your `+page.svelte.ts`, it will always have a reference that's up-to-date. The `data` prop is basically a reactive `$state()`.

Here's the `+page.svelte.ts` that'd pair good with the above `+page.ts`.
```svelte
<script lang="ts">
    import { afterNavigate, beforeNavigate } from "$app/navigation";

    let { data }: Props = $props();
</script>

<svelte:head>
    <title>Thing App | {data.thing.name}</title>
    <meta name="description" content={data.thing.description} />
</svelte:head>
```
This will update the title and description of the page on navigations, say from `/thing/1` to `/thing/2`. Unlike previous `onMount` code that only reacted on first load, here we don't even need `onMount` at all and it'll react to any `data` change from our `load()` function which again depends on the `[id]` URL parameter.

But how do you detect if that navigation happened.. or is happening? Maybe you need that functionality for extra UX. Well, that's where `beforeNavigate` & `afterNavigate` come into play, where `afterNavigate` runs after the `load` is finished (AKA, after `data` is updated), which ensures the `data` prop in your `+page.svelte` is up-to-date. Or you could just use `$effect()` to react to any changes in `data` like some people here did... but why'd you do that?

Here's another demo.
```svelte
<script lang="ts">
    import { app } from "$lib/app.svelte";
    import { afterNavigate, beforeNavigate } from "$app/navigation";

    let { data }: Props = $props();
    let is_loading_thing = $state(false);

    beforeNavigate(() => is_loading_thing = true)
    afterNavigate(() => {
        app.active_thing = data.thing; // Your global state. If you have none and your page is self-contained, just ignore this line!
        is_loading_thing = false;
    })
</script>

<svelte:head>
    <title>Thing App | {data.thing.name}</title>
    <meta name="description" content={data.thing.description} />
</svelte:head>
{#if is_loading_thing}
    <div class="loading-overlay">
        <div class="spinner"></div>
    </div>
{/if}
```

**TLDR;** Every time the `id` is invalidated such as going from `/thing/1` to `/thing/2`, this load function is naturally re-called without forcing a page reload and your code reacts to that data change the right way.

**Important:** Use `load` statelessly and don't set global states inside it. Why? Because not every `load` call is caused by a page load or user navigation. Svelte sometimes _preloads_ a page, such as when a user hovers an anchor link (`</a>`) to `/thing/...` so it'd call your `+page.ts`'s `load` function to preload that `data`. Of course you could disable that with `data-sveltekit-preload-data="off"` on anchor elements, but this still remains a valid rule nonetheless.

Example of the consquences of violating that rule: A user hovers a link... boom, the whole page changed to state related to `thing` instead of... just stay as-is. So yeah, `load` should be stateless, unless you are simply initializing app registeries / services in the root `+layout.ts` or something like that.

**Another Tip:** We talked about how to react to navigations functionally, but if for some reason you need components themselves to literally re-mount on navigations inside the same `+page.svelte`, use the `{#key}` block in Svelte to do so, which re-mounts stuff inside it when `thing.id` changes. Example:
```svelte
{#key data.thing.id}
  <ThingView thing={data.thing}/>
{/key}
```

**Comment by teemingc** [comment]
Thanks for the update @Beyondo ! I'm trying to see if we can improve the existing docs for this in https://svelte.dev/docs/kit/state-management#Component-and-page-state-is-preserved and there is a little note that says:

> If your code in onMount and onDestroy has to run again after navigation you can use [afterNavigate](https://svelte.dev/docs/kit/$app-navigation#afterNavigate) and [beforeNavigate](https://svelte.dev/docs/kit/$app-navigation#beforeNavigate) respectively.

I'm wondering if this is enough or if we should also mention `onNavigate` in this. I'm also wondering if there's a specific reason `onNavigate` was not mentioned here.

**Comment by Beyondo** [comment]
@eltigerchino I think it's because `onNavigate` is used mostly for Page/View Transitions (https://svelte.dev/docs/kit/$app-navigation#onNavigate, https://svelte.dev/blog/view-transitions). So while it can be used for setting state, it's mostly used just for animations. You could technically load state in it, but it's more _idiomatic_ to use `beforeNavigate` in case of `is_loading_thing` in my answer I suppose.

**Comment by teemingc** [comment]
Although we have that section, we can definitely do better here. Whether this means changing the default behaviour in SvelteKit 3 or making this more prominent in the docs itself such as in other pages where routes with slugs are used in examples.
